### ISSUE TITLE
SpecPrefill Phase 4 in SimpleEngine: very slow decode using `model()`

### ISSUE THREAD

**Issue by Vigilans** [waybarrios/vllm-mlx#247]
## Observed Problem

After SpecPrefill completes (scoring + sparse prefill), the Phase 4 decode loop uses manual `model()` + `mx.eval(y)` calls. On my setup, this produces **~0.3-0.4 tok/s** decode speed, while normal generation achieves 30+ tok/s.

### Resource observations

| Metric | Phase 4 decode | Normal generation |
|--------|-------------------|-------------------------------------|
| Wired memory | 30-40GB | 70-75GB |
| GPU idle | 50-65% | 0-3% |
| GPU power | <1W | 90-93W |
| Decode speed | 0.3-0.4 tok/s | 30+ tok/s |

Model weights are ~60GB (Qwen3.5-122B-A10B-4bit). During Phase 4, wired memory never reaches the full model size, and GPU is mostly idle.

## Replace with `stream_generate` fixes the decode speed

Replacing the manual decode loop with `mlx_lm.stream_generate`, passing the sparse-prefill cache via `prompt_cache`. After this change, decode speed returned to normal levels, with wired memory and GPU utilization matching normal generation behavior.

I verified with `temp=0` greedy decoding that both methods produce identical token sequences.

```python
sampler = make_sampler(temp=temperature, top_p=top_p)
first_token_id = sampler(logits[:, -1, :]).item()
first_text = tokenizer.decode([first_token_id])
eos_id = tokenizer.eos_token_id

results = [
    SimpleNamespace(
        text=first_text,
        finish_reason="stop" if first_token_id == eos_id else None,
    )
]

if first_token_id != eos_id:
    for chunk in mlx_stream_generate(
        model, tokenizer,
        prompt=mx.array([first_token_id]),
        max_tokens=max_tokens - 1,
        sampler=sampler,
        prompt_cache=cache,
    ):
        results.append(
            SimpleNamespace(
                text=chunk.text,
                finish_reason=getattr(chunk, "finish_reason", None),
            )
        )
```

Not sure if this is a proper fix. Is the difference in GPU utility expected between `model()` and `stream_generate()` ? Has anyone experienced similar issues?

## Environment

- Mac Studio M1 Ultra (64-core GPU, 128GB)
- mlx 0.31.1, mlx-lm 0.31.1
- vllm-mlx 0.2.7
- Target: Qwen3.5-122B-A10B-4bit, Draft: Qwen3.5-2B-4bit

**Comment by Thump604** [comment]
Thanks for the detailed report and diagnosis, @Vigilans. This is a real bug and your analysis is spot on.

The root cause is that the Phase 4 decode loop in `_specprefill_generate` calls `model()` directly in a plain Python loop without any pipelining. The stock `mlx_lm.stream_generate` (and the vllm-mlx engine's normal decode path) uses `mx.async_eval` to overlap Metal command buffer submission with Python-side work, and pre-fetches model weights so the GPU stays saturated. The manual loop submits one synchronous eval per token, which means the GPU idles while Python prepares the next step and vice versa. That explains the ~100x slowdown you measured.

I wrote the original SpecPrefill integration (PR #180). At the time I used a manual loop to keep the sparse KV cache handling explicit, but I should have wired it back into the engine's normal pipelined decode path after Phase 3 completes. That was an oversight.

The fix is straightforward: after Phase 4 produces the first token from the sparse-prefilled cache, hand control back to the engine's standard `stream_generate` / `_generate_step` loop for all subsequent tokens. The cache state is fully materialized at that point so there is nothing special about the remaining decode steps. Your approach of calling `stream_generate` is exactly right in principle.

I can put up a PR for this. The change would be in `_specprefill_generate` in `simple.py`: replace the manual decode loop with a handoff to the engine's existing pipelined generation, passing through the already-populated cache.

**Comment by Thump604** [comment]
I reproduced the root cause against a clean `v0.2.7` baseline. The slow path is Phase 4 decode after sparse prefill: it falls back to a manual token-by-token `model()` loop instead of handing off to `mlx_lm.stream_generate(...)`.

I now have a local fix that switches the post-prefill decode handoff to `stream_generate`, plus a regression test in my staging branch (`tests/test_simple_engine.py`) that asserts Phase 4 no longer uses manual `model()` decode.

I’m rebasing/packaging that cleanup on top of `v0.2.7` as a narrow follow-up rather than folding it into the larger batched-text work. I haven’t posted fresh throughput numbers here yet, so consider this a root-cause confirmation and fix path, not a final performance claim.

**Comment by Thump604** [comment]
Quick status update on the SimpleEngine fix and a related data point from the BatchedEngine path.

SimpleEngine fix: still in my staging branch as described above. I have not posted fresh throughput numbers yet because I want to validate against a clean v0.2.7 reproduction, not the older mixed runtime I was on previously. Tracking that as a follow-up.

Cooperative SpecPrefill on BatchedEngine: I ran a live canary on Qwen3.5-122B-A10B (4-bit) with the batched text scheduler and cooperative SpecPrefill enabled (2B draft, threshold=256, keep_pct=0.2). Results from /v1/status after a two-prompt warm-up:

- num_specprefill_requests: 1
- num_specprefill_fallbacks: 0
- text_scheduler.num_requests_processed: 2
- total_prompt_tokens: 5479, total_completion_tokens: 4
- prefix_cache: 1 hit / 1 miss, 18 tokens saved
- Metal active 88.22 GB / peak 92.23 GB

That confirms the broader SpecPrefill design is sound on BatchedEngine, and is independent of the SimpleEngine slow-decode bug this issue tracks. SimpleEngine still needs the Phase 4 stream_generate handoff fix.

**Comment by Thump604** [comment]
This issue is addressed by PR #248 (also by @Vigilans), which replaces the manual `model() + mx.eval(y)` decode loop in both `_run_specprefill` call sites with `mlx_lm.stream_generate` accepting the SpecPrefill-prepared `prompt_cache`. PR is currently MERGEABLE on main with independent technical confirmation.

**Comment by Thump604** [comment]
PR #248 fixes this by replacing the manual `model() + mx.eval()` loop with pipelined `stream_generate` handoff. The fix is approved and awaiting rebase. Will close once #248 merges.
